### PR TITLE
RFD 52/53/54: Passwordless

### PR DESCRIPTION
**Pull request by codingllama**
Design for passwordless/usernameless support in Teleport.

The design is split in various documents in order to keep them smaller and more focused. The current set of designs covers a fully-functioning local auth solution, including:

* Server-side changes
* Browser support
* FIDO2 CLI support (Linux and macOS)
* macOS CLI support for Touch ID

### CODE REVIEW

**Comment by codingllama** [comment]
FYI @klizhentas @russjones @alex-kovoy.

**Comment by codingllama** [comment]
Relevant for issues #9160 and #8499.

File: rfd/9999-passwordless-fido2.md
Comment (alex-kovoy) [inline rfd/9999-passwordless-fido2.md:122]: I do like the option (1) with all the benefits it provides. As mentioned, it gives us a free-multi-platform support which should not be underestimated. 

As for downside, I am not sure if purely FIDO2 CLI-based flow matters that much. This is because `tsh` will always require a browser for SSO which is by far the most popular Teleport auth preference. And since we know that browser UX _wins_ over CLI (with less taps), it seems like option (1) is preferable.

And another benefit is the Teleterm (upcoming native Teleport client) will be able to use `tsh` APIs as is :)

File: rfd/9999-passwordless.md
Comment (alex-kovoy) [inline rfd/9999-passwordless.md:181]: > challenge is created _without_ knowledge of the user identity.

Do we need to worry about an abuse of requests to this backend public endpoint that creates challenges for anonymous users?

File: rfd/9999-passwordless-fido2.md
Comment (codingllama) [reply rfd/9999-passwordless-fido2.md:122]: Yep, I agree with you there. It's hard to ignore the advantages of this option.

File: rfd/9999-passwordless.md
Comment (codingllama) [reply rfd/9999-passwordless.md:181]: I talk a bit about this in the Security section. 

I think, in the end, this is an inherent problem for authentication endpoints - this one is not that different. We can (and should!) rate limit access to mitigate bad behavior, but there isn't a whole lot to do other than that. (I'm open to ideas if you have any.)

Comment (alex-kovoy) [reply rfd/9999-passwordless.md:181]: I think the difference with existing public auth endpoints is that this one introduces a side affect as it creates a challenge (stored in the memory/db). So one idea would be to use a circular in-memory buffer with a reasonable size (in addition to rate limits) for storing anonymous login state.

Comment (codingllama) [reply rfd/9999-passwordless.md:181]: Noted - thanks for the suggestion. I'll add a nod to in the design later.

File: rfd/9999-passwordless-fido2.md
Comment (klizhentas) [reply rfd/9999-passwordless-fido2.md:122]: @codingllama @alex-kovoy what if we limit passwordless flow only if folks are using teleterm? Can this credentials handler used in teleport terminal?

Comment (alex-kovoy) [reply rfd/9999-passwordless-fido2.md:122]: We can support both options (1) and (2) in `teleterm`. It's just option (2) will require minor changes (when prompting for user input) which is not a big deal.

Comment (phosphore) [reply rfd/9999-passwordless-fido2.md:122]: While delegating FIDO2 authentication to browsers is attractive, won’t this “train” users to trust a browser pop-up (a thing attackers can also easily create) when starting the FIDO ceremony? Restricting this to Teleterm would be best since you would also have greater control in respect to using the default user agent for the system (through Electron it's also possible to enforce certificate pinning).

File: rfd/9999-passwordless-fido2.md
Comment (phosphore) [inline rfd/9999-passwordless-fido2.md:353]: Wouldn't this facilitate MITM attacks? It seems that the `assigneeID` secret is a single point of failure for the ceremony. What is preventing an attacker from serving their own `/web/assignee?assigneeID=n` links and phish a victim?

**Comment by phosphore** [comment]
Aside from rate-limiting the challenge endpoints, are [account locks](https://github.com/gravitational/teleport/blob/master/rfd/0009-locking.md) disabled when Passwordless authN is enabled?

**Comment by phosphore** [comment]
Is Passwordless auth required globally for all users / tenant, just like U2F/OTP? In this way an attacker can't enable FIDO passwordless authN first, locking the user out.

File: rfd/9999-passwordless-fido2.md
Comment (codingllama) [reply rfd/9999-passwordless-fido2.md:122]: I think this ship has sailed with the various Cloud CLI tools doing browser popups, but it's a fair point.

**Comment by codingllama** [comment]
> Aside from rate-limiting the challenge endpoints, are [account locks](https://github.com/gravitational/teleport/blob/master/rfd/0009-locking.md) disabled when Passwordless authN is enabled?

I expect account locks should still apply - at least I don't see why not right now. Do you have an exception in mind?

> Is Passwordless auth required globally for all users / tenant, just like it already works in the U2F/OTP design? In this way an attacker can't enable FIDO passwordless authN first, locking the user out.

The way it is designed right now, if WebAuthn is enabled then any user could register and use passwordless, but it doesn't preclude users from authentication with a "passwordfull" method.

It's also important to notice that registration requires a prior authentication ceremony, so to lock someone out in this sense an attacker would need to fully authenticate first - and at this stage the user was already fully compromised.

File: rfd/9999-passwordless-fido2.md
Comment (codingllama) [reply rfd/9999-passwordless-fido2.md:81]: We check the clone warnings in our current implementation, but I've elected to only log a warning (see [lib/auth/webauthn/login.go#L227](https://github.com/gravitational/teleport/blob/d3dca90b93f648f128c11c42ab7bcda61893ceec/lib/auth/webauthn/login.go#L227)). As you said, it's a bit of a toss whether sign counters are reliable and should be treated as a failure, but happy to change if you folks it's for the best.

(This plays a part in our WebAuthn implementation, but not so much for pwdless itself.)

File: rfd/9999-passwordless.md
Comment (codingllama) [reply rfd/9999-passwordless.md:415]: I'd be leaning on canned retrieval here, but I'll make a note in the design. Do you have any specific checks in mind, in any case?

File: rfd/9999-passwordless.md
Comment (codingllama) [reply rfd/9999-passwordless.md:445]: (1) is my personal preference, but we need actual product here. @klizhentas is the person to ask.

File: rfd/9999-passwordless-fido2.md
Comment (codingllama) [reply rfd/9999-passwordless-fido2.md:353]: I think there are a few protecting factors here:

1) `tsh` is the sole holder of the private key, fishing for the assignee ID won't give you that (the key never leaves tsh)
2) To MITM `/web/assignee` you need to impersonate the Teleport Proxy the user talks to. Assuming a proper prod setup (ie, no `tsh login --insecure`) this is hopefully hard to do.

 It's also important to note that the assignee ID is not really a secret, although it is hard to guess beforehand.

File: rfd/9999-passwordless.md
Comment (codingllama) [reply rfd/9999-passwordless.md:181]: Done.

File: rfd/9999-passwordless.md
Comment (phosphore) [reply rfd/9999-passwordless.md:415]: For the FIDO Metadata Service:
* Follow the [metadata TOC processing rules](https://fidoalliance.org/specs/fido-v2.0-rd-20180702/fido-metadata-service-v2.0-rd-20180702.html#metadata-toc-object-processing-rules) and verify the signature of the Metadata TOC object using the TOC signing certificate chain
* Check whether the status report of the authenticator model has changed compared to the cached entry (e.g. an authenticator changed from the status `REVOKED` to `FIDO_CERTIFIED` unexpectedly)
* Cross-reference the TOC returned with third party metadata (the relying party, Teleport, should obtain metadata directly from authenticator vendors or other trusted sources and cross-checks it)
* Raise a log warning if it changes from the latest cached version

For the Apple attestation statement it could be harder, but you could try to pin the TLS certificate for the endpoint if it's observed to be stable.

File: rfd/9999-passwordless-fido2.md
Comment (phosphore) [reply rfd/9999-passwordless-fido2.md:353]: I'm sorry, I could have been clearer. In the theorized attack scenario:
* An adversary `tsh` starts authentication by registering a new assignee login in Teleport, including the public key to be signed.
* The victim is tricked into completing the authentication ceremony using the resulting assignee ID in their web browser, who performs the authentication ceremony on the evil `tsh`'s behalf.
* Once the ceremony is complete, the evil `tsh` retrieves the credentials from Teleport.

Is the above scenario reasonable or am I reading the design flow incorrectly/missing something?

Comment (codingllama) [reply rfd/9999-passwordless-fido2.md:353]: An adversary `tsh` that manages to trick the user could run the login ceremony by itself as well, regardless of passwordless or assignee logins, but I think I get your point now.

What I think we could do to mitigate is to make it extra clear, in the Web UI, that we are running login on behalf of `tsh` and show all the relevant information (binary name, user, origin IP, so on, as captured by the Teleport servers).

Do you any ideas to change or secure the flow? This is an optional bit of the design at the moment, but I do foresee it being useful for Terminal.

**Comment by codingllama** [comment]
Hello folks,

I've just pushed here the first draft for [macOS Passwordless using Touch ID](https://github.com/gravitational/teleport/pull/9296/commits/81c3ba9efdf5815653c3d10cfaf81d9bd2393478).

The macOS draft completes everything necessary for local auth passwordless. At this stage we could implement a fully-functional local auth solution, so I won't add further designs to this PR (namely SSO and Windows support).

Please let me know your thoughts!

@klizhentas @russjones @phosphore

File: rfd/9999-passwordless-fido2.md
Comment (phosphore) [reply rfd/9999-passwordless-fido2.md]: I feel that we'll need to open a dedicated RFC just for this alternative, since a new threat modeling process is needed to identify the right mitigations for this flow. One of the security needs I can anticipate is the need for implementing something similar to the OAuth Proof Key for Code Exchange pattern. PKCE ([RFC 7636](https://datatracker.ietf.org/doc/html/rfc7636)) is an extension to the Authorization Code flow that would prevent certain attacks and that is needed to securely perform the OAuth exchange from public clients. The technique involves the client first creating a secret, hashing it, and then using that secret again when exchanging the authorization code for an access token. This way if the code is intercepted, it will not be useful since the token request relies on the initial secret. See https://auth0.com/docs/get-started/authentication-and-authorization-flow/authorization-code-flow-with-proof-key-for-code-exchange-pkce#how-it-works for a complete diagram of the flow. Depending on the implementation, `tsh` may need to implement a similar proof to get the credentials.

Comment (codingllama) [reply rfd/9999-passwordless-fido2.md]: Thanks for the detailed reply, Lorenzo. I'll catch up with the links and do any follow up questions later.

No objections on my part for taking this section out. We can move forward without it and cross the Terminal bridge when we have to (maybe they can move to "pure" Web authentication or get away with using `tsh` flows).

**Comment by codingllama** [comment]
FYI @r0mant @zmb3

**Comment by codingllama** [comment]
Hey @nklaassen, could I bother you for a review as well? Thanks!

**Comment by codingllama** [comment]
I've just pushed a few commits removing the "browser-based CLI login" alternative and adding a bunch more details in the UX sections. No changes in the engineering design.

FYI @xinding33 for UX feedback.

**Comment by codingllama** [comment]
(Btw, one of the commit names looks off, going to amend it now.)

File: rfd/9999-passwordless-fido2.md
Comment (codingllama) [reply rfd/9999-passwordless-fido2.md]: Removed section from the design. We can revisit in a future design, if at all.

File: rfd/0053-passwordless-fido2.md
Comment (nklaassen) [inline rfd/0053-passwordless-fido2.md:160]: what do these credentials look like, are they readable enough that it's not completely unreasonable to make the user pick?

File: rfd/0053-passwordless-fido2.md
Comment (nklaassen) [inline rfd/0053-passwordless-fido2.md:260]: I'm guessing there will be a unique RPID per teleport cluster? Would it ever make sense for a user to have multiple credentials on the same hardware key for Teleport? If not, is there any way for us to prevent them from doing that?

File: rfd/0053-passwordless-fido2.md
Comment (nklaassen) [inline rfd/0053-passwordless-fido2.md:310]: Would `tsh` need any special permissions to access hardware keys on any of the supported OS? If any, can they be requested at runtime? I'm hoping we don't need to run as root on linux or with admin rights on Windows, or something like that.

ninja edit - I see this is covered in the next doc for MacOS. Do we need to do anything similar for Linux or Windows?

File: rfd/0053-passwordless-fido2.md
Comment (nklaassen) [inline rfd/0053-passwordless-fido2.md:410]:
```suggestion
This alternative seems a bit extreme, at least for the moment, so it wasn't
```

**Review by nklaassen — COMMENTED** [review]
How does this interact with SAML/SSO logins, is it an alternative for local users only?

**Comment by codingllama** [comment]
@nklaassen 

> How does this interact with SAML/SSO logins, is it an alternative for local users only?

It doesn't (for now), the designs here are local-only.

File: rfd/0053-passwordless-fido2.md
Comment (codingllama) [reply rfd/0053-passwordless-fido2.md:160]: They have human-readable field specifically for this purpose (ie the Teleport username), so should be fine.

https://www.w3.org/TR/webauthn-2/#dictionary-user-credential-params

File: rfd/0053-passwordless-fido2.md
Comment (codingllama) [reply rfd/0053-passwordless-fido2.md:260]: Correct, we have a unique RPID per cluster. It's a similar situation to the old U2F AppID.

They may have credentials for multiple Teleport users (probably rare for the average user, but I do it all the time). It's fine if they do, WebAuthn foresees that, plus I don't see any problems in allowing it.

File: rfd/0053-passwordless-fido2.md
Comment (codingllama) [reply rfd/0053-passwordless-fido2.md:310]: Nope, we don't need special permissions for FIDO2 - in fact it's the exact same thing we do today for U2F. That is why you should be wary of apps interacting with your Yubikeys - they could be pretending to be anyone.

MacOS / Touch ID is a different beast, in general it's much safer in terms of permissions and app isolation, but one should still be careful with whom they trust.

File: rfd/0053-passwordless-fido2.md
Comment (nklaassen) [reply rfd/0053-passwordless-fido2.md:260]: I meant, can a single Teleport user have multiple credentials on the same hardware key? IE both credentials reference the same Teleport user in the backend. I could see someone configuring that by mistake (maybe while trying to add a second/backup key) and getting confused by seeing multiple logins with the same name. Probably not a big deal though, they could just delete duplicates

**Review by nklaassen — APPROVED** [review]

File: rfd/0054-passwordless-macos.md
Comment (klizhentas) [inline rfd/0054-passwordless-macos.md:179]: @codingllama any reason that can't be set by the `auth-preferences` by admin?

File: rfd/0054-passwordless-macos.md
Comment (klizhentas) [inline rfd/0054-passwordless-macos.md:190]: Can we make auto choice here? For MacOS make this `TouchID` for Linux, make this `FIDO2` and remove the flag?

File: rfd/9999-passwordless-fido2.md
Comment (klizhentas) [reply rfd/9999-passwordless-fido2.md]: @codingllama for `Teleport Terminal`, as it's an electron app, can we simply use the `web` flow using `credentials provider` without any pop ups?

File: rfd/0053-passwordless-fido2.md
Comment (codingllama) [reply rfd/0053-passwordless-fido2.md:260]: Oh, got it. [We don't allow that for WebAuthn](https://github.com/gravitational/teleport/blob/bd01152781443609fad53ab52722c9c38785eb53/lib/auth/webauthn/register.go#L138-L153), so should be safe in this front.

File: rfd/0052-passwordless.md
Comment (codingllama) [reply rfd/0052-passwordless.md:444]: Easy enough to do.

I'll update the design to say we are going with a canned set of roots for passwordless, but we drop those as soon as the admin defines any attestation lists.

File: rfd/0052-passwordless.md
Comment (codingllama) [reply rfd/0052-passwordless.md:462]: Done. Spelled out mitigations more clearly in the design, including a (simple) IP rate limiter and a circular in-memory buffer for challenges.

File: rfd/0054-passwordless-macos.md
Comment (codingllama) [reply rfd/0054-passwordless-macos.md:190]: The reasoning for the `--mfa_mode` flag is similar to https://github.com/gravitational/teleport/pull/9296#discussion_r797730723 - `auto` locks the user on Touch ID, but they may want to use a FIDO2 key for some reason, so this exists to allow that. We could make the flag hidden, but as in the other thread, I do find it useful to have.

That said, I would expect most users to rely on the default/auto behavior and never touch this.

File: rfd/9999-passwordless.md
Comment (codingllama) [reply rfd/9999-passwordless.md]: Picked (2) / shipping with a canned set of roots, based on Sasha's feedback (see https://github.com/gravitational/teleport/pull/9296#discussion_r797232387).

The actual roots are TBD and I don't plan on specifying them in the design (that is more of a fluid product decision).

**Comment by codingllama** [comment]
Pushed https://github.com/gravitational/teleport/pull/9296/commits/c331f80bd6b2c3f8d9adf95e52077169348f6c99, documenting the need for a macOS .app for `tsh`. This closes the last outstanding TODO.

File: rfd/0052-passwordless.md
Comment (codingllama) [reply rfd/0052-passwordless.md:188]: Thanks for pointing it out, done.

File: rfd/0054-passwordless-macos.md
Comment (zmb3) [inline rfd/0054-passwordless-macos.md:247]: Great idea!

**Review by zmb3 — APPROVED** [review]

File: rfd/0053-passwordless-fido2.md
Comment (codingllama) [reply rfd/0053-passwordless-fido2.md:353]: For FIDO2 I think so. The best we can tell is if there are any hardware keys plugged in the system, from this point on we would have to assume that one of them holds a passwordless key and start the flow. If we guessed wrong there won't be a resident key and we'll fail partway, with I think at least one needless touch, so not that great.

Touch ID makes this a lot nicer, since in that case we can tell if a key exists without user interaction.

File: rfd/0054-passwordless-macos.md
Comment (codingllama) [reply rfd/0054-passwordless-macos.md:247]: I like my debug commands. :)

This one is borne of paranoia that a client somewhere will fail under mysterious reasons, so there's at least some data we can ask them for.

**Review by phosphore — APPROVED** [review]
With the latest additions the RFD design looks solid to me, good job @codingllama!

**Review by xinding33 — COMMENTED** [review]
Overall looks great. Only significant comments are around the UX for `TouchID` in the case where multiple users exist for a `TouchID` credential.

File: rfd/0053-passwordless-fido2.md
Comment (codingllama) [reply rfd/0053-passwordless-fido2.md:392]: libfido2 does that for us, so it's not that bad. If libfido2 happens to fail for a particular device, then there isn't much we can do regardless.

PIN registration flows usually have retries built-in (eg, in Chrome). I'm not planning to do that - we try PIN+confirmation once, if it fails then the user can run the command again. (Up for discussion, of course, but the simpler it is, lesser the chance I do something wrong.)

File: rfd/0054-passwordless-macos.md
Comment (codingllama) [reply rfd/0054-passwordless-macos.md:201]: That was actually my first draft, more below.

> From a security perspective, we show the TouchID user all available users tied to their TouchID credential. That seems fine though because all these users do belong to the TouchID user.

The information leakage is what prompted me to change it to the more "obfuscated" version. I'm not concerned about the legitimate user, but if a third party that gets their hands on a (unlocked) macOS they could see which credentials it holds by doing `tsh login --proxy=example.com` and cancelling after the prompt. It's not huge, but I'd rather sidestep it if I can.

File: rfd/0054-passwordless-macos.md
Comment (codingllama) [reply rfd/0054-passwordless-macos.md:302]: Definitely, if there is less ambiguity than all the better.  👍

File: rfd/0053-passwordless-fido2.md
Comment (xinding33) [reply rfd/0053-passwordless-fido2.md:392]: Thanks for the clarification!

File: rfd/0054-passwordless-macos.md
Comment (xinding33) [reply rfd/0054-passwordless-macos.md:201]: Understood. The risk, from a UX perspective, is that the user might not recall the exact login they want to use. What are some of the easiest ways that a user can recall their login? Perhaps reverse-i-search and checking the web app?

**Review by xinding33 — APPROVED** [review]
I expect the `Touch ID` multi-user use case to be fairly uncommon and you're right in that it presents some information leakage. I'm good with this version of the RFD! Nice job, it was a fun and educational read!

**Comment by codingllama** [comment]
Hey folks, we have Security, UX and Eng approvals; plus @klizhentas also made a pass, so I think we're good. Going to hit the merge button now. If you have any further comments let me know, I'll be happy to open a PR to address them.

Many thanks for all the thoughtful feedback, y'all.

**Pull request merged**
